### PR TITLE
[Mellanox] CMIS host management script

### PR DESCRIPTION
**Pull request by skr31**
#### Why I did it
Added a script for CMIS Host Management enabling and disabling.
- Prevents bugs caused by wrong configuration of the feature
- More convenient for costumers to configure the feature 

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Add the script to /usr/bin/ that supports two options:
--enable <path> <path> - receives paths to config files for Port SI parameters and Module SI parameters: 
                                             media_settings.json, optics_si_settings.json, and enables the feature
--disable - disables the feature

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2021-2024 NVIDIA CORPORATION & AFFILIATES.
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: please change the year to be only 2024

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        CMISHostMgmtActivator.change_param("pmon_daemon_control", sku_path, 'enable')
+        CMISHostMgmtActivator.change_param("sai_xml", sku_path, 'enable')
+        
+
+@click.command()
+@click.option('--disable', is_flag=True, help='Disable CMIS Host Management')
+@click.option('--enable', nargs=2, type=click.Path(), help='Enable CMIS Host Management, receives two arguments: media_settings.json path, and optics_si_settings.json path')
+def main(disable, enable):
+
+    import pdb; pdb.set_trace()
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: please remove PDB related lines

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    if disable and enable:
+        print("Error: can't use both options, please choose one.")
+
+    if disable:
+        CMISHostMgmtActivator.disable_im()
+
+    elif enable:
+        CMISHostMgmtActivator.enable_im(enable)
+
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py:166]: If user didn't provide disable/enable flag, maybe it's better to print to the screen that nothing was executed. 
Do we have "force" to one of the parameters in "click"?

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            "disabled_param": "<late-create-all-ports>1</late-create-all-ports>" # Shouldn't be called
+        }
+    }
+
+    @staticmethod
+    def change_param(param, path, action):
+        file_path = '{}/{}'.format(path, CMISHostMgmtActivator.PARAMS[param]["file_name"])
+        lines = None
+
+        with open(file_path, 'r') as param_file:
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: why not to open the file from the beginning as 'rw'?

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        sku_path = '/usr/share/sonic/device/{0}/{1}'.format(platform, sku)
+        platform_path = '/usr/share/sonic/device/{0}'.format(platform)
+        CMISHostMgmtActivator.change_param("sai_profile", sku_path, 'disable')
+        CMISHostMgmtActivator.change_param("pmon_daemon_control", platform_path, 'disable')
+        
+        CMISHostMgmtActivator.remove_im_file('{0}/{1}'.format(sku_path, 'media_settings.json'))
+        CMISHostMgmtActivator.remove_im_file('{0}/{1}'.format(sku_path,'optics_si_settings.json'))
+        CMISHostMgmtActivator.remove_im_file('{0}/{1}'.format(platform_path, 'media_settings.json'))
+        CMISHostMgmtActivator.remove_im_file('{0}/{1}'.format(platform_path, 'optics_si_settings.json'))
+        CMISHostMgmtActivator.remove_im_file('{0}/{1}'.format(sku_path, 'pmon_daemon_control.json'))
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: what will happen if pmon_daemon_control.json will be only under SKU folder? 
and then user wants to disable the feature and you will remove it? 
I would suggest to check - if there is such file under platform folder, remove the file under SKU and change the platform one param. if there is no such file, change param in the file under SKU

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                platform = value
+
+            elif key == 'HwSKU':
+                sku = value
+
+        return platform, sku
+
+
+    @staticmethod
+    def remove_im_file(file_path):
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: "im" stands for "independent module"?
maybe change it just to remove_file/copy_file?

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        CMISHostMgmtActivator.remove_im_file('{0}/{1}'.format(platform_path, 'optics_si_settings.json'))
+        CMISHostMgmtActivator.remove_im_file('{0}/{1}'.format(sku_path, 'pmon_daemon_control.json'))
+
+
+    @staticmethod
+    def enable_im(args):
+        platform, sku = CMISHostMgmtActivator.parse_show_platform_summary()
+        sku_path = '/usr/share/sonic/device/{0}/{1}'.format(platform, sku)
+        platform_path = '/usr/share/sonic/device/{0}'.format(platform)
+
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py:132]: @skr31 @dprital
do we want to have some kind of platform limitation? 
Or not, so we won't have to update the script each time we add support to a new platform? 

My suggestion is at least not approve the enablement of the feature is the platform is lower that spc3.
If the user tries to do it, need to print some error and return.

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        },
+        "pmon_daemon_control": {
+            "file_name": "pmon_daemon_control.json",
+            "enabled_param": "\"skip_xcvrd_cmis_mgr\": false",
+            "disabled_param": "\"skip_xcvrd_cmis_mgr\": true",
+        },
+        "sai_xml": {
+            "file_name": "sai_<>.xml", # will be filled at main, since we can't know the SKU here 
+            "enabled_param": "<late-create-all-ports>1</late-create-all-ports>",
+            "disabled_param": "<late-create-all-ports>1</late-create-all-ports>" # Shouldn't be called
```
Comment (noaOrMlnx) [inline platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: Indeed it should not be called. must we have it as part of this structure? 
If so, maybe just put some empty string? ""

**Review by noaOrMlnx — CHANGES_REQUESTED** [review]

**Comment by noaOrMlnx** [comment]
@skr31 
Please add in description, that the config files are for Port SI parameters and Module SI parameters and their names - media_settings.json, optics_si_settings.json.
Thanks

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 1 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: Fixed

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 2 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: Fixed

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 3 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py:166]: Added print

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 6 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: Changed

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 4 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: Fixed

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 8 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: Fixed

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 5 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py]: Fixed

File: platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py (hunk 7 above)
Comment (skr31) [reply platform/mellanox/cmis_host_mgmt/cmis_host_mgmt.py:132]: Added check for spc3 and higher for now

**Review by noaOrMlnx — APPROVED** [review]

**Pull request closed without merging**